src/AeroSharp/DataAccess/MapAccess/Generators/MapEntryGeneratorWithSerializer.cs

```diff
@@ -0,0 +1,20 @@
+﻿using AeroSharp.Serialization;
+using Aerospike.Client;
+
+namespace AeroSharp.DataAccess.MapAccess.Generators;
+
+/// <summary>
+///     A <see cref="IMapEntryGenerator"/> implementation that serializes the map value before wrapping
+///     it in a <see cref="Value"/>. This is useful when the key is a simple scalar type and the map value
+///     is a complex type that needs to be serialized.
+/// </summary>
+internal sealed class MapEntryGeneratorWithSerializer : IMapEntryGenerator
+{
+    private readonly ISerializer _valueSerializer;
+
+    public MapEntryGeneratorWithSerializer(ISerializer valueSerializer) => _valueSerializer = valueSerializer;
+
+    public Value GenerateKey<TKey>(TKey key) => Value.Get(key);
```

Comment:
I guess it depends on how serialized this becomes, but the documentation was pretty clear I thought on what keys can be:

Map keys can be of type String, Bytes, Integer or Double.

So in this case, @farmerau, are you suggesting adding the ability to serialize a key?  I think a large part of the performance gain is just sticking to those keys and not adding more complexity at first here.

Reply:
My concern is we have no type constraint on `TKey`.

I’m not sure what `Value.Get` is going to return for a non immediately convertible type, but I’m hoping it’s not the HashCode.

So it’d be good to know ahead of time what the behavior might be (it might be covered in the tests I haven’t looked at yet) and more importantly that the keys are deterministic for arbitrary (value equivalent) objects.

Reply:
I actually had support for opt-in key serialization to allow for arbitrarily complex keys, including associated tests, but ripped it out since I wasn't sure there would be a use case for it. Based on this convo it sounds like it might be nice to include.

I can add that back in, or do type checking of `TKey` within the builder to enforce simple keys. Thoughts?

Reply:
I think simple key enforcement is probably leading the user to the happiest path. I realize we can't do "OR" type constraints (bummer, but makes sense) so we'd need to handle it at runtime.

I think offering complex objects could be user friendly but might lead folks down a path of worse-than-necessary performance.

Reply:
I went ahead and added this key type checking, which verifies that the key type must be one of the supported Aerospike scalar data types: https://docs.aerospike.com/server/guide/data-types/scalar-data-types. 

Let me know if this is what you were imagining, or if you think there's a better way of doing this.

Reply:
Initially, I was going to propose something like this:

Since we know all of the types at compile-time, we could take advantage of C# pattern matching. It would be very similar to what the Aerospike client library does [here](https://github.com/aerospike/aerospike-client-csharp/blob/master/AerospikeClient/Value/Value.cs#L235C9-L259).

Something silly like:

```csharp
object integer = 3;
switch(integer) {
    case int myInt:
    case double myDouble:
    default:
        throw new Exception();
}
```

However, I think it really depends on whether or not we think `MapBuilder` might be used in a hot path or not. I think, in all cases, pattern matching would reduce memory allocations but then you lose the convenience of the valid types string you throw in the exception. In a hot path (let's say someone is putting the creation of these into a scoped injection lifetime), it could matter. This does lose the `Type.IsAssignableFrom` check, but that may be okay.

I also think that, based on their documentation, we should include support for `ulong` since they support both unsigned and signed 64 bit integers.

Reply:
Hmm, @farmerau are you thinking we'd remove the type validation in the builder and add this switch statement here in the `MapEntryGenerator.GenerateKey` method, throwing an exception when an unexpected key type is encountered? 

My only hesitation would be that we're deferring the type validation from a spot that _might_ be in the hot path to a spot that is much more likely to be in the hot path, but maybe the pattern matching benefit of reducing memory allocations makes that negligible. 🤔

Reply:
Oh, no, I was thinking something like replacing these lines [here](https://github.com/wbaldoumas/AeroSharp/blob/48e8bdf990cd438e6af0f20315decd921a0e6156/src/AeroSharp/MapBuilder.cs#L138-L143):

```csharp
if (!ValidKeyTypes.Any(validKeyType => validKeyType.IsAssignableFrom(keyType)))
{
    var validKeyNames = string.Join(", ", ValidKeyTypes.Select(type => $" {type}"));

    throw new UnsupportedKeyTypeException($"Map keys must be of type{validKeyNames}.");
}
```

Reply:
We can keep it in this maybe-hot-path but do more to reduce it's impact.

Reply:
Ah, what would that look like, then? We've only got a type parameter `TKey` to work there in the builder, and not a variable to pass into a  `switch` statement that uses pattern matching.

Reply:
One way I could see improving this is making the valid types a `HashSet<Type>` and doing `.Contains(...)` to make this an ~O(1) operation, rather than using the `.Any(...)`.

```csharp
private static readonly HashSet<Type> ValidKeyTypes = new HashSet<Type>
{
    typeof(long), typeof(double), typeof(string), typeof(byte[]), typeof(bool), typeof(ulong)
};
```

```csharp
private IMapOperator<TKey, TValue> BuildMapOperator<TKey, TValue>()
{
    var keyType = typeof(TKey);

    if (!ValidKeyTypes.Contains(keyType))
    {
        var validKeyNames = string.Join(", ", ValidKeyTypes.Select(type => $"{type}"));

        throw new UnsupportedKeyTypeException($"Map keys must be of type {validKeyNames}.");
    }

    // ...
}
```

Reply:
Oh, you’re totally right. I got lost in my example and didn’t think about the details surrounding it.

I think the HashSet solution is optimal, though!

Reply:
Is it possible for this to go through and open an issue to add the validation in an upcoming PR?  

I do agree that making sure that our keys are appropriately set is important, but for v.0.0.1 of this Map support, I think users would _hopefully_ read the clear documentation on AS's site about support?

Reply:
I've got a PR right on the heels of this that I'd be happy to add to

Reply:
Let me clean this up to use the `HashSet` way of validating the types and then we should be able to merge this in. 

Note that I'm not including a version bump yet or releasing a new version of the package, as we may want to refine this further and add additional map methods before we consider a release, so I'm fine with you adjusting this more in a follow up PR, @awesor.

That said, @farmerau I gave it a go using some `ulong` keys within the integration tests at very high values and I'm seeing them get converted to `-1` on the Aerospike server when read back out. I don't actually think unsigned is supported.

Reply:
Cool beans - that's really interesting with ulong there!

Reply:
@wbaldoumas Weird on `ulong`-- their docs made me think it was supported. Thanks for testing!

Reply:
Alright, I pushed up that quick change to swap to a `HashSet` to hold our valid types. Barring any objections, I'll merge this in tomorrow night and hand it on over to you, @awesor, with your follow up PR. Looking forward to our upcoming release! 😄

Reply:
Also, looks like that last push dismissed both of your reviews, so this will need another stamp of approval from both of you. Thanks for all the feedback!